### ISSUE TITLE
Support control of accessibility-related features in the same way

### ISSUE THREAD

**Issue by zepumph** [phetsims/chipper#1031]
For features like pan/zoom, description, and voicing, we are finding that we want to control these in similar ways. We have a package.json flag that can be overwritten by a queryParameter. There are a variety of ways to handle this currently, but we want to refactor all to use the pattern that pan/zoom does, in which it calls upon the packageJSON value as the default value for the query parameter:

https://github.com/phetsims/chipper/blob/9df5f3519c14254ef7c68080976cb95a4492023c/js/initialize-globals.js#L632-L643

On top of this refactor, we think there are now enought features to specify these inside their own, new object called `features`, like so:

```json
      "phet": {
        "colorProfile": true,
        "requirejsNamespace": "AREA_MODEL_DECIMALS",
        "phetLibs": [
          "area-model-common",
          "twixt",
          "vegas"
        ],
        "runnable": true,
        "supportedBrands": [
          "phet",
          "adapted-from-phet"
        ],
        "simulation": true,
        "features": {
          "supportsInteractiveDescription":true,
          "supportsVoicing": true,
          "supportsGestureControl": true,
          "supportsPanAndZoom": true
        }
      }
```

@jessegreenberg and I will implement for the above mentioned packagejson flags (supportsVoicing doesn't currently exist).

We will then reevaluate and work with JB to see if he wants to update the pattern for controlling sound and enhanced sound.

In overview:
- [ ] Turn all feature-related query parameters into booleans with a default value coming from package.json
- [ ] Refactor package.json into "features" section with all the above properties.

**Comment by zepumph** [comment]
Looks like https://github.com/phetsims/rutherford-scattering/commit/1a2c916a7cdbc030173e41cd68d1aa2c9cc702a3 needed to have grunt update called on it.

**Comment by zepumph** [comment]
In the above commits, I made some progress. It is easy to do the simple ones, but there will be many more changes for voicing and description. I have completed the move/refactor of `supportsGestureControl` and `supportsPanAndZoom` to be under `features`. This removed `Sim.supportsPanAndZoom`.

Up next:
- [ ] Convert `?supportsDescription` to boolean
- [ ] Convert `?supportsVoicing` to boolean
- [ ] Move `supportsInteractiveDescription in package to be under `phet.features`
- [ ] Move `supportsVoicing` in package to be under `phet.features`

**Comment by zepumph** [comment]
While there are half of the flags buried inside "features", and half out, this should be high priority.

**Comment by zepumph** [comment]
Right now I'm feeling like waiting until the `preferences` branch is merged into master before working on `supportsVoicing`. I may still make the `?supportsInteractiveDescription` changes before then.

**Comment by zepumph** [comment]
Above I did 4 things with interactive description control:
* Converted the query parameter to a boolean
* renamed to `supportsInteractiveDescription`
* Refactored the query parameter so that its default value could be taken from the package flag
* moved the package flag to nest under `features`

I also created https://github.com/phetsims/chipper/issues/1038 to work on `supportsVoicing`. In general things are looking really good. I'm happy with this pattern and didn't run into problems.

------------------

To @jbphet:

The only thing I see left for this issue is to convert `supports(Enhanced)Sound` to work the same way. I think it would be best to discuss this with @jbphet at some point. In general, the proposal is for a consistent feature set, and to run features from query parameters instead of sim flags. `supportsInteractiveDescription` is the best example of this at this time:
* It can be set by default in `packageJSON.phet.features.supportsInteractiveDescription` (default to falsey)
* The query parameter reads from that value by default. 
* Everywhere that accesses the feature uses `queryParameters.supportsInteractiveDescription`.

I would like to convert sound related feature flags too to be consistent and as clear as possible. Let me know if that would work and I'd be happy to do the work.

**Comment by jbphet** [comment]
I'm generally okay with the idea of `supportsSound` and `supportsEnhancedSound` being moved to the `features` section in `package.json`, and with there being query parameters to override the values in `package.json`, but there are some things that will need to be addressed as part of the effort:
- You'll need to figure out how you want to handle illegal configurations, such as having `supportsSound` false and `supportsEnhancedSound` true.  It should probably assert during development, but do we want to throw an error if done by users on built versions via the query params?
- There is already a `sound` query parameter that is public and may be in use by users outside of PhET.  We will need to decide if we want to change this to make it more orthogonal to the new parameters, which may be to make it into `soundInitiallyEnabled` or something.  I don't know whether users outside of PhET are currently using the `sound` query parameter, so we should check in with people who _do_ know (@kathy-phet maybe?) to help decide whether it's okay to change this.
- There is already an `enhancedSoundInitiallyEnabled` that should probably work the same way it does now, but perhaps we need to check for illegal configurations such as this being true when `supportsEnhancedSound` is false.

**Comment by zepumph** [comment]
>You'll need to figure out how you want to handle illegal configurations, such as having supportsSound false and supportsEnhancedSound true. It should probably assert during development, but do we want to throw an error if done by users on built versions via the query params?

I feel like sound is always just a subset of enhanced sound, and that is also how Joist feels currently:

https://github.com/phetsims/joist/blob/6c5612088a367ee312dc231ee076afee4f3d9d12/js/Sim.js#L502-L507

I think being graceful, and allowing all possible parameters without error is the nicest and safest. Isn't this already technically an issue currently if you were to add `?supportsEnhancedSound` to a runtime in which supportsSound was false? 

>There is already an enhancedSoundInitiallyEnabled that should probably work the same way it does now, but perhaps we need to check for illegal configurations such as this being true when supportsEnhancedSound is false.

Again, I feel like this is a problem now also. Is there an error currently if you add the flag `enhancedSoundInitiallyEnabled` without enhanced sound enabled?


>There is already a sound query parameter that is public and may be in use by users outside of PhET. We will need to decide if we want to change this to make it more orthogonal to the new parameters, which may be to make it into soundInitiallyEnabled or something. I don't know whether users outside of PhET are currently using the sound query parameter, so we should check in with people who do know (@kathy-phet maybe?) to help decide whether it's okay to change this.

I think that is public, and used outside of phet. Let's not change this one, and it doesn't need to come from the packageJSON anyways, so I think we are good to go here.

**Comment by zepumph** [comment]
I converted supportsEnhancedSound to a boolean, and nested it under package.json's "features" object. I tested it with all possible combinations in GFL, and it looked go, so I committed above. Next will be `?supportsSound`.

**Comment by zepumph** [comment]
I added support for supportsSound above.


>There is already an enhancedSoundInitiallyEnabled that should probably work the same way it does now, but perhaps we need to check for illegal configurations such as this being true when supportsEnhancedSound is false.

I see that this query parameter controls a booleanProperty passed as an enabledControlProperty to the sound manager. Thus this too is a graceful output, not changed by this issues work (since it behaves the same as it did prior to the change), and is working as expected.

I'm ready to close this, issue, and to keep query parameters graceful like they have been. @jbphet anything else here?

**Comment by jessegreenberg** [comment]
Over slack @jbphet noticed that the sound disappeared from the navigation bar and we thought it was probably related to this change. It looks like some references to `packageJSON.phet.supportsSound` were not updated so the Sim thinks that sound is not supported.

I also noticed a few places where `supportsSound` was not moved under "features" in package.json, in area-builder, arithmetic, faradays-law, and gravity-force-lab-basics.

**Comment by jessegreenberg** [comment]
Hmm, I see https://github.com/phetsims/arithmetic/commit/bd47db88297990f40bb3458af2c02710cc475d4c for arithemtic and https://github.com/phetsims/area-builder/commit/91ff00eef86047fd86e6443351bdcb749668d5c1 for area-builder. But in the above commits I moved supportsSound under features in gravity-force-lab-basics and faradays-law

Then I updated the usages of packageJSON related to supportsSound. @zepumph can you please review?

**Comment by zepumph** [comment]
Sorry! It looks like commits weren't pushed even when I inspected the webstorm dialog. Thanks for taking the quick lead for the fix. I pushed my commits from last week above. Note that you don't need packageJSON from Sim.js because that info in now held by the query parameter.

**Comment by zepumph** [comment]
Ready for real review now, over to @jbphet.

**Comment by jbphet** [comment]
The idea and general approach seems great, as do most of the code changes.  Two comments:

1. The key name `features` seems awfully generic to me.  Can we be a bit more specific?  Are this accessibility features?  Optional features?  Add ons?
2. I tested the sound-oriented query parameters, since that's what I'm most familiar with.  I noticed that using the parameters `&supportsSound=false&supportsEnhancedSound=false` with John Travoltage cause the speaker icon to disappear from the nav bar, but I think it's still supposed to be there to allow muting of voice.  I'm not sure how this is supposed to behave, and there might be other issues already created for this, but I thought I'd mention it so that @jessegreenberg can make sure this case is covered.

**Comment by zepumph** [comment]
>The key name features seems awfully generic to me. Can we be a bit more specific? Are this accessibility features? Optional features? Add ons?

@jbphet and @jessegreenberg, how about `simFeatures`?


>I tested the sound-oriented query parameters, since that's what I'm most familiar with. I noticed that using the parameters &supportsSound=false&supportsEnhancedSound=false with John Travoltage cause the speaker icon to disappear from the nav bar, but I think it's still supposed to be there to allow muting of voice.

@jessegreenberg, is this something that is going to be handled by https://github.com/phetsims/joist/issues/724?

**Comment by zepumph** [comment]
renamed to simFeatures above (and below). @jbphet, anything else here in your opinion? Feel free to close.

**Comment by jbphet** [comment]
Looks good, closing.
